[14.0][MIG] l10n_es_intrastat_report: Migration to 14.0

joao-p-marques opened the pull request:
@Tecnativa
TT28080

ping @pedrobaeza @victoralmau

luc-demeyer committed:
add intrastat module

danielduqma committed:
[11.0][ADD] l10n_es_intrastat_product_declaration

Commit:
[MIG] l10n_es_intrastat_product: Migration to 12.0

pedrobaeza committed:
[MIG+IMP+FIX] l10n_es_intrastat_report: Finish migration to 12 + several imps/fixes

* Remove unneeded dependency to stock_picking_invoice_link.
* Updated HS Codes source file.
* Optimized compute/import methods.
* Remove useless/obsolete code, data and fields.
* Added Spanish translation.

pedrobaeza committed:
[FIX] l10n_es_intrastat_report: Allow invoices with no linked purchases

Previous code was forcing the existence of linked purchase lines which may be not
true for all. As we already have a fallback in declaration code, let's not force
this at location level.

joao-p-marques committed:
[IMP] l10n_es_intrastat_report: black, isort, prettier

joao-p-marques committed:
[MIG] l10n_es_intrastat_report: Migration to 13.0

TT27865

joao-p-marques committed:
[IMP] l10n_es_intrastat_report: black, isort, prettier

victoralmau commented on l10n_es_intrastat_report/models/l10n_es_intrastat_product_declaration.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class L10nEsIntrastatProductComputationLine(models.Model):
+    _name = "l10n.es.intrastat.product.computation.line"
```
Please add `_description` field to prevent warning log

victoralmau commented on l10n_es_intrastat_report/models/l10n_es_intrastat_product_declaration.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class L10nEsIntrastatProductDeclarationLine(models.Model):
+    _name = "l10n.es.intrastat.product.declaration.line"
```
Please add `_description` field to prevent warning log

victoralmau commented on l10n_es_intrastat_report/wizards/l10n_es_intrastat_code_import.py:
```diff
@@ -0,0 +1,97 @@
+
+
+class L10nEsPartnerImportWizard(models.TransientModel):
+    _name = "l10n.es.intrastat.code.import"
```
Please add `_description` field to prevent warning log

victoralmau commented on l10n_es_intrastat_report/security/l10n_es_intrastat_report_security.xml:
```diff
@@ -6,6 +6,6 @@
         <field
             name="domain_force"
-        >[('company_id', 'child_of', user.company_id.ids)]</field>
+        >[('company_id', 'child_of', user.company_ids.ids)]</field>
```
Although it's work fine i don't really sure that it's totally correct.

pedrobaeza replied:
This should be updated in v13:

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0

joao-p-marques replied:
Yes, I missed that in the migration. Making the change now and then I'll rebase here.

victoralmau replied:
These changes it's merged in 13.0, please rebase.

joao-p-marques committed:
[FIX] l10n_es_intrastat_report: Update multi-company rules

Missed step from the migration to v13
Update multi-company security domains and provide migration scripts

TT27865

joao-p-marques commented:
Cambios hechos @victoralmau @pedrobaeza

pedrobaeza commented on l10n_es_intrastat_report/tests/test_l10n_es_intrastat_report.py:
```diff
@@ -62,6 +71,21 @@ def setUpClass(cls):
 
+    def test_post_init_hook(self):
+        post_init_hook(self.env.cr, None)
+        items = self.env["ir.model.data"].search(
```
Esto no es correcto, porque está dependiendo del estado de la BD, que puede incluso no tener el plan contable español puesto. Lo que puedes hacer para testearlo es:

- Crear una posición fiscal a mano con intrastat forzado a False.
- Crear el XML-ID a mano (ir.model.data). Ponle como name `l10n_es_intrastat_report_fp_intra_private` para evitar que choque con cualquier XML-ID existente, y como de todas formas el dominio lo incluye, no hay problema.
- Lanza el hook.
- Comprueba en tu posición fiscal que intrastat es True.

joao-p-marques replied:
@pedrobaeza ya he hecho los cambios. Gracias!

pedrobaeza commented:
Travis en rojo

joao-p-marques commented:
Creo que el último cambio ya lo corrige. Es necesaria la librería `xlrd` en la versión 1.2.0 para importar de ficheros Excel. [Odoo la instala por defecto si Python está en la versión 3.8](https://github.com/odoo/odoo/blob/14.0/requirements.txt#L56), pero como parece que Travis está corriendo con otra versión, así nos aseguramos.

pedrobaeza commented:
Travis está en rojo

joao-p-marques commented:
> Travis está en rojo

Están fallando ahora porque no consigue instalar la versión correcta de la librería por conflicto de dependencias... Lo he añadido ahora a `test-requirements.txt` que me parece que lo usa Travis para instalar las dependencias. Si no, quizás @Yajo pueda echar una mano aquí.

yajo commented:
El error parece ser el típico fallo de Geonames: https://travis-ci.com/github/OCA/l10n-spain/jobs/488415061#L1707

Efectivamente, si visito http://download.geonames.org/export/zip/ES.zip me sale:

![imagen](https://user-images.githubusercontent.com/973709/110086523-63e76500-7d8a-11eb-83c3-bbd8d49cdf55.png)

🤷🏼‍♂️

pedrobaeza commented:
Pero no termino de entender por qué necesita esa versión de la librería xlrd y no vale con cualquiera.

joao-p-marques commented:
@pedrobaeza acabo de subir un cambio para que use un fichero `.xls` en lugar de `.xlsx`, que estaba causando problemas. Así, creo que no dará el error de antes, aunque use la versión más antigua de la dependencia. Parece que no quieren tampoco mantener suporte a estos ficheros en la biblioteca: https://github.com/python-excel/xlrd#xlrd
Gracias @Yajo

joao-p-marques commented:
Bueno, están fallando ahora por Geonames, pero creo que de mi parte los cambios están completos.

pedrobaeza commented on l10n_es_intrastat_report/tests/test_l10n_es_intrastat_report.py:
```diff
@@ -59,9 +75,41 @@ def setUpClass(cls):
+        )
+        post_init_hook(self.env.cr, None)
+        items = self.env["ir.model.data"].search(
+            [
+                ("model", "=", "account.fiscal.position"),
+                ("name", "like", "fp_intra_private"),
+                ("module", "=", "l10n_es"),
+            ]
+        )
+        self.assertTrue(item in items)
+        intra_private_fps = self.env["account.fiscal.position"].browse(
+            items.mapped("res_id")
+        )
+        self.assertTrue(fp in intra_private_fps)
+        for fp in intra_private_fps:
+            self.assertTrue(fp.intrastat)
```
Todo esto sobra ya. Solo es necesario `self.assertTrue(fp.intrastat)`.

pedrobaeza commented on l10n_es_intrastat_report/wizards/l10n_es_intrastat_code_import.py:
```diff
@@ -34,7 +35,7 @@ def _import_hs_codes(self):
             raise exceptions.UserError(_("xlrd library not found."))
         code_obj = self.env["hs.code"]
         path = os.path.join(
```
Aquí utiliza `get_resource_path` para mayor fiabilidad.

pedrobaeza reviewed (changes requested):
Falta también un archivo data que ponga en el account.fiscal.position.template de l10n_es.fp_intra_private el flag de instrastat a True para nuevas compañías que se creen a partir de ese momento.

pedrobaeza commented:
@joao-p-marques puedes rematar esto?

joao-p-marques committed:
[MIG] l10n_es_intrastat_report: Migration to 14.0

joao-p-marques commented:
@pedrobaeza cambios hechos :+1:

pedrobaeza commented:
Travis está en rojo

joao-p-marques commented:
> Travis está en rojo

Están fallando por `l10n_es_aeat_sii_oca`. Creo que esta rama estaba desactualizada, así que he hecho rebase de la principal y así vemos si era eso.

joao-p-marques commented:
Ya está en verde @pedrobaeza

pedrobaeza commented:
/ocabot merge nobump
